File: services/valuation-engine/valuation_engine/edgar/companyfacts.py
"""
Fetch and lightly structure the XBRL "companyfacts" bundle for a company.

Endpoint: https://data.sec.gov/api/xbrl/companyfacts/CIK##########.json
Shape:    facts -> taxonomy ("us-gaap" | "dei") -> tag -> units -> unit -> [fact, ...]
Each fact: {end, val, accn, fy, fp, form, filed, frame?, start?}
"""
from __future__ import annotations

from dataclasses import dataclass
from datetime import date

from .client import EdgarClient
from .tickers import CompanyRef

COMPANYFACTS_URL = "https://data.sec.gov/api/xbrl/companyfacts/CIK{cik}.json"


@dataclass(frozen=True)
class Fact:
    taxonomy: str
    tag: str
    unit: str
    value: float
    end: date
    start: date | None
    form: str
    fp: str | None
    fy: int | None
    accession: str
    filed: date
    frame: str | None

    @property
    def duration_days(self) -> int | None:
        if self.start is None:
            return None
        return (self.end - self.start).days

    @property
    def is_instant(self) -> bool:
        return self.start is None


@dataclass
class CompanyFacts:
    ref: CompanyRef
    entity_name: str
    facts: dict[str, list[Fact]]  # key: "taxonomy:tag"

    def get(self, taxonomy: str, tag: str) -> list[Fact]:
        return self.facts.get(f"{taxonomy}:{tag}", [])


def _parse_date(s: str | None) -> date | None:
    return date.fromisoformat(s) if s else None


async def fetch_companyfacts(client: EdgarClient, ref: CompanyRef) -> CompanyFacts:
    raw = await client.get_json(COMPANYFACTS_URL.format(cik=ref.cik_padded))
    return parse_companyfacts(raw, ref)


def parse_companyfacts(raw: dict, ref: CompanyRef) -> CompanyFacts:
    """Pure function so tests can feed recorded fixtures without the network."""
    facts: dict[str, list[Fact]] = {}
    for taxonomy, tags in raw.get("facts", {}).items():
        for tag, body in tags.items():
            key = f"{taxonomy}:{tag}"
            bucket: list[Fact] = []
            for unit, rows in body.get("units", {}).items():
                for row in rows:
                    try:
                        bucket.append(
                            Fact(
                                taxonomy=taxonomy,
                                tag=tag,
                                unit=unit,
                                value=float(row["val"]),
                                end=date.fromisoformat(row["end"]),
                                start=_parse_date(row.get("start")),
                                form=row.get("form", ""),
                                fp=row.get("fp"),
                                fy=row.get("fy"),
                                accession=row.get("accn", ""),
                                filed=date.fromisoformat(row["filed"]),
                                frame=row.get("frame"),
                            )
                        )
                    except (KeyError, ValueError, TypeError):
                        continue
            if bucket:
                facts[key] = bucket
    return CompanyFacts(ref=ref, entity_name=raw.get("entityName", ref.name), facts=facts)

File: packages/valuation-core/src/commonMain/kotlin/com/swcsoftware/valuelens/core/CompanyFacts.kt
package com.swcsoftware.valuelens.core

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

class CompanyRefCore(val ticker: String, val cik: Long, val name: String) {
    val cikPadded: String get() = cik.toString().padStart(10, '0')
}

/** One XBRL fact from SEC's companyfacts bundle. */
class Fact(
    val taxonomy: String, val tag: String, val unit: String, val value: Double,
    val end: Day, val start: Day?, val form: String, val fp: String?, val fy: Int?,
    val accession: String, val filed: Day, val frame: String?,
) {
    val durationDays: Long? get() = start?.daysUntil(end)
    val isInstant: Boolean get() = start == null
}

class CompanyFacts(val ref: CompanyRefCore, val entityName: String, private val facts: Map<String, List<Fact>>) {
    fun get(taxonomy: String, tag: String): List<Fact> = facts["$taxonomy:$tag"] ?: emptyList()
    val tagKeys: Set<String> get() = facts.keys
    /** Facts under a "taxonomy:tag" key — used by coverage to judge candidate tags. */
    fun factsFor(key: String): List<Fact> = facts[key] ?: emptyList()
}

object CompanyFactsParser {
    private val json = Json { ignoreUnknownKeys = true }

    /** Pure parse of the companyfacts JSON text (same structure the Python reference reads). */
    fun parse(text: String, ref: CompanyRefCore): CompanyFacts {
        val root = json.parseToJsonElement(text).jsonObject
        val entityName = root["entityName"]?.jsonPrimitive?.content ?: ref.name
        val out = LinkedHashMap<String, MutableList<Fact>>()
        val facts = root["facts"]?.jsonObject ?: JsonObject(emptyMap())
        for ((taxonomy, tags) in facts) {
            for ((tag, body) in tags.jsonObject) {
                val units = body.jsonObject["units"]?.jsonObject ?: continue
                val bucket = mutableListOf<Fact>()
                for ((unit, rows) in units) {
                    for (row in rows.jsonArray) {
                        val o = row.jsonObject
                        val value = o["val"]?.jsonPrimitive?.doubleOrNull ?: continue
                        val end = Day.parse(o["end"]?.jsonPrimitive?.content) ?: continue
                        val filed = Day.parse(o["filed"]?.jsonPrimitive?.content) ?: continue
                        bucket += Fact(
                            taxonomy, tag, unit, value, end, Day.parse(o["start"]?.jsonPrimitive?.content),
                            o["form"]?.jsonPrimitive?.content ?: "", o["fp"]?.jsonPrimitive?.content,
                            o["fy"]?.jsonPrimitive?.intOrNull, o["accn"]?.jsonPrimitive?.content ?: "", filed,
                            o["frame"]?.jsonPrimitive?.content,
                        )
                    }
                }
                if (bucket.isNotEmpty()) out["$taxonomy:$tag"] = bucket
            }
        }
        return CompanyFacts(ref, entityName, out)
    }
}
